core/http-auth/src/main/java/software/amazon/awssdk/http/auth/internal/DefaultAwsV4HttpSigner.java

```diff
@@ -72,157 +75,225 @@ public SyncSignedRequest sign(SyncSignRequest<? extends AwsCredentialsIdentity>
+        if (payload != null) {
+            payload.subscribe(bodyDigester);
+        }
 
-            SdkHttpRequest.Builder builder = doSign(request,
-                new ContentChecksum(digestHex, sdkChecksum));
+        return bodyDigester.digestBytes().thenApply(BinaryUtils::toHex);
+    }
 
-            return builder.build();
-        });
+    /**
+     * Generate a {@link CanonicalRequestV2} from the {@link SignRequest},the {@link SdkHttpRequest},
+     * and the {@link ContentChecksum}.
+     */
+    protected CanonicalRequestV2 createCanonicalRequest(SignRequest<?, ? extends AwsCredentialsIdentity> signRequest,
+                                                        SdkHttpRequest request,
+                                                        ContentChecksum contentChecksum) {
+        boolean doubleUrlEncode = validatedProperty(signRequest, DOUBLE_URL_ENCODE, true);
+        boolean normalizePath = validatedProperty(signRequest, NORMALIZE_PATH, true);
+
+        return new CanonicalRequestV2(request, contentChecksum.contentHash(), new CanonicalRequestV2.Options(
+            doubleUrlEncode,
+            normalizePath
+        ));
+    }
 
-        return AsyncSignedRequest.builder()
-            .request(CompletableFutureUtils.joinLikeSync(signedReqFuture))
-            .payload(request.payload().orElse(null))
-            .build();
+    /**
+     * Generate a string-to-sign using the algorithm, the {@link CredentialScope}, and the hash of the canonical request.
+     */
+    protected String createSignString(String algorithm, CredentialScope credentialScope, String canonicalRequestHash) {
+        LOG.debug(() -> "AWS4 Canonical Request Hash: " + canonicalRequestHash);
+
+        String stringToSign = algorithm +
+            SignerConstant.LINE_SEPARATOR +
+            credentialScope.getDatetime() +
+            SignerConstant.LINE_SEPARATOR +
+            credentialScope.scope() +
+            SignerConstant.LINE_SEPARATOR +
+            canonicalRequestHash;
+
+        LOG.debug(() -> "AWS4 String to sign: " + stringToSign);
+        return stringToSign;
     }
 
     /**
-     * Adds session credentials to the request given.
+     * Generate a signature using the string-to-sign and the signing key.
      */
-    private void addSessionCredentials(SdkHttpRequest.Builder requestBuilder,
-                                       AwsSessionCredentialsIdentity credentials) {
-        requestBuilder.putHeader(SignerConstant.X_AMZ_SECURITY_TOKEN, credentials.sessionToken());
+    protected String createSignature(String stringToSign, byte[] signingKey) {
+        return BinaryUtils.toHex(
+            SignerUtils.computeSignature(stringToSign, signingKey)
+        );
+    }
+
+    /**
+     * Add the signature to the request in some form (as a header, query-parameter, or otherwise).
+     */
+    protected void addSignature(SdkHttpRequest.Builder requestBuilder, String algorithm, AwsCredentialsIdentity credentials,
```

Comment:
nit: formatting inconsistency. some params are in same line, some are in different lines.

Reply:
Hmm... I wish check-style would enforce this kind of consistency.